src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php

```diff
@@ -73,4 +75,90 @@ protected function processValue($value, $isRoot = false)
 
         return $value;
     }
+
+    /**
+     * @param Definition $definition
+     * @param bool       $required
+     *
+     * @return \ReflectionFunctionAbstract|null
+     *
+     * @throws RuntimeException
+     */
+    protected function getConstructor(Definition $definition, $required)
+    {
+        if (is_string($factory = $definition->getFactory())) {
+            if (!function_exists($factory)) {
+                throw new RuntimeException(sprintf('Unable to resolve service "%s": function "%s" does not exist.', $this->currentId, $factory));
+            }
+            $r = new \ReflectionFunction($factory);
+            if (false !== $r->getFileName() && file_exists($r->getFileName())) {
+                $this->container->fileExists($r->getFileName());
+            }
+
+            return $r;
+        }
+
+        if ($factory) {
+            list($class, $method) = $factory;
+            if ($class instanceof Reference) {
+                $class = $this->container->findDefinition((string) $class)->getClass();
+            } elseif (null === $class) {
+                $class = $definition->getClass();
+            }
+            if ('__construct' === $method) {
```

Comment:
Ok, good point too :) It should still be added in `CheckDefinitionValidityPass` imo though.